File: sdk/python/src/dstack_sdk/verify_env_encrypt_public_key.py
# SPDX-FileCopyrightText: © 2025 Phala Network <dstack@phala.network>
#
# SPDX-License-Identifier: Apache-2.0

"""Verify ECDSA signatures on KMS env-encrypt public keys.

The KMS signs the X25519 env-encrypt public key it returns from
``/GetAppEnvEncryptPubKey`` so deployers can prove the key originated from a
specific signer before encrypting secrets against it. There are two message
formats:

- ``signature_v1`` (preferred): includes a Unix timestamp to bound replay.
- legacy: pubkey + app_id only. Kept for backward compatibility with old KMS
  builds. Vulnerable to replay; use the v1 variant whenever available.

Both formats sign ``keccak256(prefix + b":" + app_id + [timestamp_be_bytes] +
public_key)`` with secp256k1, producing a 65-byte ``r || s || recovery_id``
signature. This module recovers the signer's compressed public key.
"""

import time
from typing import Optional
import warnings

from eth_keys import keys
from eth_utils import keccak

DEFAULT_MAX_AGE_SECONDS = 300
_PREFIX = b"dstack-env-encrypt-pubkey"
_SEPARATOR = b":"
_FUTURE_SKEW_TOLERANCE_SECONDS = 60


def _normalize_app_id(app_id: str) -> Optional[bytes]:
    if app_id.startswith("0x"):
        app_id = app_id[2:]
    try:
        return bytes.fromhex(app_id)
    except ValueError:
        return None


def _recover_signer(msg_hash: bytes, signature: bytes) -> Optional[str]:
    try:
        recovered = keys.Signature(
            signature_bytes=signature
        ).recover_public_key_from_msg_hash(msg_hash)
        return "0x" + recovered.to_compressed_bytes().hex()
    except Exception:
        return None


def verify_env_encrypt_public_key(
    public_key: bytes,
    signature: bytes,
    app_id: str,
    timestamp: int,
    *,
    max_age_seconds: int = DEFAULT_MAX_AGE_SECONDS,
) -> Optional[str]:
    """Verify a timestamp-protected KMS env-encrypt public key signature.

    Returns the signer's compressed secp256k1 public key (0x-prefixed hex) on
    success, or ``None`` on bad signature length, expired/future timestamp,
    invalid hex app_id, or signature recovery failure.
    """
    if len(signature) != 65:
        return None

    now = int(time.time())
    age = now - timestamp
    if age < -_FUTURE_SKEW_TOLERANCE_SECONDS:
        return None
    if age > max_age_seconds:
        return None

    app_id_bytes = _normalize_app_id(app_id)
    if app_id_bytes is None:
        return None

    timestamp_bytes = timestamp.to_bytes(8, "big")
    message = _PREFIX + _SEPARATOR + app_id_bytes + timestamp_bytes + public_key
    return _recover_signer(keccak(message), signature)


def verify_env_encrypt_public_key_legacy(
    public_key: bytes,
    signature: bytes,
    app_id: str,
) -> Optional[str]:
    """Verify a legacy (non-timestamped) KMS signature.

    .. deprecated::
        Legacy signatures do not protect against replay attacks. Use
        :func:`verify_env_encrypt_public_key` with a timestamp from the KMS
        response whenever possible.
    """
    warnings.warn(
        "verify_env_encrypt_public_key_legacy is deprecated; use "
        "verify_env_encrypt_public_key with a timestamp from KMS instead.",
        DeprecationWarning,
        stacklevel=2,
    )
    if len(signature) != 65:
        return None

    app_id_bytes = _normalize_app_id(app_id)
    if app_id_bytes is None:
        return None

    message = _PREFIX + _SEPARATOR + app_id_bytes + public_key
    return _recover_signer(keccak(message), signature)

File: sdk/js/src/verify-env-encrypt-public-key.ts
// SPDX-FileCopyrightText: © 2025 Phala Network <dstack@phala.network>
//
// SPDX-License-Identifier: Apache-2.0

import { keccak_256 } from "@noble/hashes/sha3"
import { secp256k1 } from "@noble/curves/secp256k1"

const DEFAULT_MAX_AGE_SECONDS = 300

export interface VerifyOptions {
  /** Maximum age of the signed response in seconds. Default: 300 (5 minutes). */
  maxAgeSeconds?: number
}

function bigintToBeBytes(value: bigint, length: number): Uint8Array {
  const bytes = new Uint8Array(length)
  for (let i = length - 1; i >= 0; i--) {
    bytes[i] = Number(value & 0xffn)
    value >>= 8n
  }
  return bytes
}

function hexToBytes(hex: string): Uint8Array | null {
  if (hex.startsWith("0x") || hex.startsWith("0X")) hex = hex.slice(2)
  if (hex.length % 2 !== 0 || !/^[0-9a-fA-F]*$/.test(hex)) return null
  const bytes = new Uint8Array(hex.length / 2)
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.substr(i, 2), 16)
  }
  return bytes
}

function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("")
}

function concat(...parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((n, p) => n + p.length, 0)
  const out = new Uint8Array(total)
  let offset = 0
  for (const part of parts) {
    out.set(part, offset)
    offset += part.length
  }
  return out
}

function recoverSigner(
  messageHash: Uint8Array,
  signature: Uint8Array,
): string | null {
  try {
    const sigBytes = signature.slice(0, 64)
    const recovery = signature[64]
    const recoveredPubKey = secp256k1.Signature.fromCompact(sigBytes)
      .addRecoveryBit(recovery)
      .recoverPublicKey(messageHash)
    return "0x" + bytesToHex(recoveredPubKey.toRawBytes(true))
  } catch (error) {
    console.error("signature verification failed:", error)
    return null
  }
}

/**
 * Verify a timestamp-protected KMS env-encrypt public key signature.
 *
 * Returns the signer's compressed secp256k1 public key on success, or `null`
 * on failure (bad length, expired timestamp, invalid signature).
 */
export function verifyEnvEncryptPublicKey(
  publicKey: Uint8Array,
  signature: Uint8Array,
  appId: string,
  timestamp: bigint | number,
  options?: VerifyOptions,
): string | null {
  if (signature.length !== 65) return null

  const ts = typeof timestamp === "bigint" ? timestamp : BigInt(timestamp)
  const maxAge = options?.maxAgeSeconds ?? DEFAULT_MAX_AGE_SECONDS
  const now = BigInt(Math.floor(Date.now() / 1000))
  const age = now - ts
  if (age < -60n) {
    console.error("timestamp is too far in the future")
    return null
  }
  if (age > BigInt(maxAge)) {
    console.error(`timestamp is too old: ${age}s > ${maxAge}s`)
    return null
  }

  const appIdBytes = hexToBytes(appId)
  if (!appIdBytes) return null

  const prefix = new TextEncoder().encode("dstack-env-encrypt-pubkey")
  const separator = new TextEncoder().encode(":")
  const timestampBytes = bigintToBeBytes(ts, 8)
  const message = concat(
    prefix,
    separator,
    appIdBytes,
    timestampBytes,
    publicKey,
  )
  return recoverSigner(keccak_256(message), signature)
}

/**
 * @deprecated Use {@link verifyEnvEncryptPublicKey} with timestamp. Legacy
 * signatures do not protect against replay attacks.
 */
export function verifyEnvEncryptPublicKeyLegacy(
  publicKey: Uint8Array,
  signature: Uint8Array,
  appId: string,
): string | null {
  if (signature.length !== 65) return null

  const appIdBytes = hexToBytes(appId)
  if (!appIdBytes) return null

  const prefix = new TextEncoder().encode("dstack-env-encrypt-pubkey")
  const separator = new TextEncoder().encode(":")
  const message = concat(prefix, separator, appIdBytes, publicKey)
  return recoverSigner(keccak_256(message), signature)
}
